core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -461,81 +459,81 @@ public Path getWorkingDirectory() {
+        throw new IOException(String.format("Authority \"%s\" is unknown. The client can not be "
+                + "configured with the authority from %s", auth, uri));
+      }
 
-    mUri = URI.create(mAlluxioHeader);
+      mUri = URI.create(mAlluxioHeader);
+      Map<String, Object> uriConfProperties = getConfigurationFromUri(uri);
 
-    Map<String, Object> uriConfProperties = getConfigurationFromUri(uri);
+      AlluxioProperties alluxioProps = ConfigurationUtils.defaults();
+      mergeConfigurations(uriConfProperties, conf, alluxioProps);
+      AlluxioConfiguration alluxioConf = new InstancedConfiguration(alluxioProps);
 
-    synchronized (INIT_LOCK) {
-      if (sInitialized) {
-        if (!connectDetailsMatch(uriConfProperties, conf)) {
-          LOG.warn(ExceptionMessage.DIFFERENT_CONNECTION_DETAILS.getMessage(
-              FileSystemContext.get().getMasterInquireClient().getConnectDetails()));
-          initializeInternal(uriConfProperties, conf);
-        }
+      Subject subject = getHadoopSubject();
+      if (subject != null) {
+        LOG.debug("Using Hadoop subject: {}", subject);
       } else {
-        initializeInternal(uriConfProperties, conf);
+        LOG.debug("No Hadoop subject. Using FileSystem Context without subject.");
       }
-      // Must happen inside the lock so that the global filesystem context isn't changed by a
+
+      LOG.info("Initializing filesystem context with connect details {}",
+          Factory.getConnectDetails(alluxioConf));
+      mFsContext = FileSystemContext.create(subject, alluxioConf);
+
+      // Sets the file system.
+      //
+      // Must happen inside the lock so that the filesystem context isn't changed by a
       // concurrent call to initialize.
-      updateFileSystemAndContext();
-      sInitialized = true;
+      mFileSystem = FileSystem.Factory.get(mFsContext);
+      mInitialized = true;
     }
   }
 
   /**
-   * Initializes the default contexts if the connection details specified in the URI + hadoop conf
-   * is different from the default one.
+   * Merges the URI configuration with the Hadoop and Alluxio configuration, returning an
+   * {@link AlluxioConfiguration} with all properties merged into one object.
    *
    * @param uriConfProperties the configuration properties from the input uri
    * @param conf the hadoop conf
+   * @param alluxioProps Alluxio configuration properties
    */
-  void initializeInternal(Map<String, Object> uriConfProperties,
-      org.apache.hadoop.conf.Configuration conf) throws IOException {
-    // Load Alluxio configuration if any and merge to the one in Alluxio file system. These
-    // modifications to ClientContext are global, affecting all Alluxio clients in this JVM.
-    // We assume here that all clients use the same configuration.
-    HadoopConfigurationUtils.mergeHadoopConfiguration(conf, Configuration.global());
+  private void mergeConfigurations(Map<String, Object> uriConfProperties,
```

Comment:
can we actually return a new instanced conf here, instead of updating one in place?